portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,33 +14,95 @@
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content and handles data. */
+/**
+ * Servlet that handles input comments data and returns the contents
+ * with the limit number of maximum comments' number sent by viewers.
+ */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private List<String> commentsStorage = new ArrayList<>();
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = new Gson().toJson(commentsStorage);
+    Query query = new Query("Comment");
+    Iterable<Entity> results = (datastore.prepare(query)).asIterable();
+    List<String> commentsStorage = new ArrayList<>();
+    List<Integer> boundsStorage = new ArrayList<>();
+
+    // Collect all comments and the preferred maximum numbers of comments.
+    results.forEach(e -> {
+      addCommentContents(e, commentsStorage);
+      getMaxCommentsNum(e, boundsStorage);
+    });
+
+    // Show the specific number of comments.
+    String json = new Gson().toJson(collectShownComments(commentsStorage, boundsStorage));
     response.setContentType("text/html");
     response.getWriter().println(json);
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String comment = request.getParameter("comments-input");
-    commentsStorage.add(comment);
+    String maxCommentsNumString = request.getParameter("max-comments-num");
+    int maxCommentsNum = 0;
+    if (maxCommentsNumString != null) {
+      maxCommentsNum = Integer.parseInt(maxCommentsNumString);
+    }
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("contents", comment);
+    commentEntity.setProperty("bound", maxCommentsNum);
+    datastore.put(commentEntity);
     response.setContentType("text/html;");
     response.getWriter().println("Thank you for your comments!");
-    response.sendRedirect("/index.html");
+  }
+
+  /** For a specific entity, add the inserted maximum comments' number to the collection. */
+  public void getMaxCommentsNum(Entity entity, List<Integer> boundsStorage) {
+    if (entity.getProperty("bound") != null) {
+      long boundLong = (long) entity.getProperty("bound");
+      int bound = Math.toIntExact(boundLong);
+      boundsStorage.add(bound);
+    }
+  }
+
+  /** For a specific entity, add its comment's contents to the storage. */
+  public void addCommentContents(Entity entity, List<String> commentsStorage) {
+    String contents = (String) entity.getProperty("contents");
+    if (contents != "") {
+      commentsStorage.add(contents);
+    }
+  }
+
+  /** Controls the output comments' collection with the maximum comments' number limitation. */
```

Comment:
'Controls' is not very descriptive. Could this method be described better?

Reply:
Yes. I changed it to "Returns the output comments' collection with ..." because this method returns the list for "commentsStorage" finally.